guidance/llms/_openai.py

```diff
@@ -147,15 +147,20 @@ def __init__(self, model=None, caching=True, max_retries=5, max_calls_per_min=60
         if deployment_id is None:
             deployment_id = os.environ.get("OPENAI_DEPLOYMENT_ID", None)
 
-        # auto detect chat completion mode
+        # Auto-detect whether to use the chat completion mode based on the model name
         if chat_mode == "auto":
-            # parse to determin if the model need to use the chat completion API
-            chat_model_pattern = r'^(gpt-3\.5-turbo|gpt-4)(-\d+k)?(-\d{4})?$'
-            if re.match(chat_model_pattern, model):
-                chat_mode = True
-            else:
-                chat_mode = False
-        
+            # Regular expression pattern to match both standard and fine-tuned model names
+            # Examples of matched model names:
+            # - 'gpt-3.5-turbo'
+            # - 'gpt-4'
+            # - 'ft:gpt-3.5-turbo:my-org:custom_suffix:id'
+            # - 'ft:gpt-4:my-org:custom_suffix'
+            # - 'ft:gpt-3.5-turbo-1106:personal::id'
+            chat_model_pattern = r'^(ft:)?(gpt-3\.5-turbo|gpt-4)(-\d+)?(:\w+)?(::\w+)?$'
```

Comment:
Won't match new gpt4 turbo:
gpt-4-1106-preview

try something like:
^(ft:)?(gpt-3\.5-turbo|gpt-4)(-\d+)?(:\w+)?(::\w+)?.*$

Reply:
Thanks @adamgordonbell for pointing this out. I've updated the regex to accommodate this specific format, while ensuring flexibility for similar future model naming variations. The revised pattern has been thoroughly tested against all provided examples. Looking forward to any further feedback.

Reply:
Great! ( I don't have ability to merge these changes, just noting the issue )

I think this PR has a solid approach to this issue as well. Does the API used here return finetuned models?

https://github.com/guidance-ai/guidance/pull/420/files#diff-ff829083f9a27b1841e9109bbecf5648881da28e1cb4e9ddd27046233fdb0d3cR152

Reply:
Thanks for sharing the PR link, @adamgordonbell. It's an interesting approach, but the `openai.Model.list()` method unfortunately doesn't include fine-tuned models, which in my case, are essential for my thesis project. I agree that combining openai.Model.list() with a regex fallback for fine-tuned models could be a great solution. Any assistance in implementing this, especially to handle fine-tuned models, would be greatly appreciated.